tests/components/homekit_controller/test_config_flow.py

```diff
@@ -103,678 +97,402 @@ def test_valid_pairing_codes(pairing_code):
-    }
+    service = accessory.add_service(ServicesTypes.LIGHTBULB)
+    on_char = service.add_char(CharacteristicsTypes.ON)
+    on_char.value = 0
 
-    flow = _setup_flow_handler(hass)
+    bridge.add_accessory(accessory)
 
-    finish_pairing = asynctest.CoroutineMock()
+    return controller.add_device(bridge)
 
-    discovery = mock.Mock()
-    discovery.start_pairing = asynctest.CoroutineMock(return_value=finish_pairing)
 
-    flow.controller.find_ip_by_device_id = asynctest.CoroutineMock(
-        return_value=discovery
-    )
+@pytest.mark.parametrize("upper_case_props", [True, False])
+@pytest.mark.parametrize("missing_csharp", [True, False])
+async def test_discovery_works(hass, controller, upper_case_props, missing_csharp):
+    """Test a device being discovered."""
+    device = setup_mock_accessory(controller)
+    discovery_info = get_device_discovery_info(device, upper_case_props, missing_csharp)
 
     # Device is discovered
-    result = await flow.async_step_zeroconf(discovery_info)
+    result = await hass.config_entries.flow.async_init(
+        "homekit_controller", context={"source": "zeroconf"}, data=discovery_info
+    )
     assert result["type"] == "form"
     assert result["step_id"] == "pair"
-    assert flow.context == {
+    assert get_flow_context(hass, result) == {
         "hkid": "00:00:00:00:00:00",
+        "source": "zeroconf",
         "title_placeholders": {"name": "TestDevice"},
         "unique_id": "00:00:00:00:00:00",
     }
 
     # User initiates pairing - device enters pairing mode and displays code
-    result = await flow.async_step_pair({})
+    result = await hass.config_entries.flow.async_configure(result["flow_id"])
     assert result["type"] == "form"
     assert result["step_id"] == "pair"
-    assert discovery.start_pairing.call_count == 1
-
-    pairing = mock.Mock(pairing_data={"AccessoryPairingID": "00:00:00:00:00:00"})
-
-    pairing.list_accessories_and_characteristics = asynctest.CoroutineMock(
-        return_value=[
-            {
-                "aid": 1,
-                "services": [
-                    {
-                        "characteristics": [
-                            {"type": "23", "value": "Koogeek-LS1-20833F"}
-                        ],
-                        "type": "3e",
-                    }
-                ],
-            }
-        ]
-    )
 
-    finish_pairing.return_value = pairing
-
-    flow.controller.pairings = {"00:00:00:00:00:00": pairing}
-
-    result = await flow.async_step_pair({"pairing_code": "111-22-333"})
+    # Pairing doesn't error error and pairing results
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={"pairing_code": "111-22-333"}
+    )
     assert result["type"] == "create_entry"
     assert result["title"] == "Koogeek-LS1-20833F"
-    assert result["data"] == pairing.pairing_data
+    assert result["data"] == {}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5066052)